### PR TITLE
sequencer: Use compact.Range directly

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
The `compact.Range` type provides a superset of what `compact.Tree` can do.
This change removes dependency from the `compact.Tree` type in Trillian, and
allows deleting it completely.

Addresses #1338

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

File: log/sequencer.go [hunk 1]
```diff
@@ -416,11 +456,15 @@ func (s Sequencer) IntegrateBatch(ctx context.Context, tree *trillian.Tree, limi
 		stageStart = s.timeSource.Now()
 
 		// Create the log root ready for signing
-		seqTreeSize.Set(float64(merkleTree.Size()), label)
+		rootHash, err := merkleTree.GetRootHash()
```
Comment (pavelkalinnikov) [inline log/sequencer.go]: Gotcha: `rootHash` can be nil if the tree is empty. Probably won't happen, but we should better double check it here.

File: log/sequencer.go [hunk 2]
```diff
@@ -167,7 +169,19 @@ func (s Sequencer) initMerkleTreeFromStorage(ctx context.Context, root *types.Lo
...
 	}
 
-	return compact.NewTreeWithState(s.hasher, root.TreeSize, hashes, root.RootHash)
+	rng, err := fact.NewRange(0, root.TreeSize, hashes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create compact.Range: %v", err)
+	}
+	hash, err := rng.GetRootHash(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to restore the root hash: %v", err)
```
Comment (Martin2112) [inline log/sequencer.go]: Not sure about "restore" here. Can maybe pass the error through directly.

File: log/sequencer.go [hunk 3]
```diff
@@ -211,25 +225,28 @@ func (s Sequencer) prepareLeaves(leaves []*trillian.LogLeaf, begin uint64, label
 
 // updateCompactTree adds the passed in leaves to the compact tree. Returns a
 // map of all updated tree nodes, and the new root hash.
-func (s Sequencer) updateCompactTree(mt *compact.Tree, leaves []*trillian.LogLeaf, label string) (map[compact.NodeID][]byte, []byte, error) {
+func (s Sequencer) updateCompactTree(mt *compact.Range, leaves []*trillian.LogLeaf, label string) (map[compact.NodeID][]byte, []byte, error) {
```
Comment (Martin2112) [inline log/sequencer.go]: May already be on the list but the name of this func is now out of date.

File: log/sequencer.go (hunk 2 above)
Comment (pavelkalinnikov) [reply log/sequencer.go]: Changed to "compute".

File: log/sequencer.go (hunk 3 above)
Comment (pavelkalinnikov) [reply log/sequencer.go]: Done.

**Pull request merged**